[openweathermap] Add support for persisting OneCall API forecasts

florian-h05 opened the pull request:
Implement time series support introduced by https://github.com/openhab/openhab-core/pull/3597 for all forecast channels of the OneCall API weather and forecast Thing.

Also change the behaviour of the handler, so that setting `forecastMinutes`, `forecastHours` and `forecastDays` to 0 does not exclude the respective data from the API response. So these config parameters can be used to disable the individual channel groups for each time-step.

florian-h05 commented:
@jlaur I guess I need to use update instructions here to make those new channel groups available for "old" Things. Can you probably provide some help here since I have never worked with update instructions before?

jlaur commented:
> @jlaur I guess I need to use update instructions here to make those new channel groups available for "old" Things. Can you probably provide some help here since I have never worked with update instructions before?

You could have a look at https://www.openhab.org/docs/developer/bindings/thing-xml.html#updating-thing-types and: https://github.com/openhab/openhab-addons/blob/a4da9863301dec962ec4cf395bce813761ed24f7/bundles/org.openhab.binding.danfossairunit/src/main/resources/OH-INF/update/instructions.xml#L12-L20

florian-h05 commented:
Thanks.

Is there a way to add a whole channel group in one step? Or do I have to add each individual channel?

florian-h05 commented:
@jlaur I got the Thing update instructions working, so from my POV this is ready for merging.

florian-h05 commented on bundles/org.openhab.binding.openweathermap/src/main/java/org/openhab/binding/openweathermap/internal/handler/OpenWeatherMapOneCallHandler.java:
```diff
@@ -512,7 +497,123 @@ private void updateHourlyForecastChannel(ChannelUID channelUID, int count) {
+        }
+        if (channelId.equals(CHANNEL_TIME_STAMP) || channelId.equals(CHANNEL_CONDITION)
+                || channelId.equals(CHANNEL_CONDITION_ID) || channelId.equals(CHANNEL_CONDITION_ICON)
+                || channelId.equals(CHANNEL_CONDITION_ICON_ID)) {
```
Note that I have decided to not create TimeSeries for non-Number channels, because obviously you cannot create a diagram for those.
However having those as TimeSeries data might still be useful for rules and scripts, I guess.

@jlaur WDYT?

florian-h05 replied:
@jlaur Gentle ping, as IMHO this should make it into 4.1.
Would be great if there is more than one add-on (EnergiDataService) supporting the new time-series feature.

jlaur replied:
I guess in order to fully replace the old work-around of having an enormous amount of channels to represent values at different points in time, you would need to include those non-number channels as well?

For reference: https://www.openhab.org/addons/bindings/openweathermap/#one-call-api-minutely-forecast

This is only about persistence, and as you mention, rules can also take advantage of this, not only charts. You could also access the data in a database from something completely outside openHAB.

Perhaps @J-N-K has an opinion on this?

florian-h05 replied:
> I guess in order to fully replace the old work-around of having an enormous amount of channels to represent values at different points in time, you would need to include those non-number channels as well?

Agreed. I will also enable those channels (I hope to get to this tomorrow, now is a bit too late.)

florian-h05 replied:
Implemented in https://github.com/openhab/openhab-addons/pull/15963/commits/32f404cb90f25659a24cf151abd53cb15b74a7bc.

jlaur commented on bundles/org.openhab.binding.openweathermap/src/main/resources/OH-INF/thing/channel-types.xml:
```diff
@@ -169,6 +169,14 @@
 	</channel-group-type>
 
+	<channel-group-type id="oneCallMinutelyTimeSeries">
+		<label>One Call API Minutely Forecast Time Series</label>
```
I'm wondering if you need to include "time series" in the label. States will be pushed to the event bus as they become current, so perhaps this should just be considered having a time dimension. I think this new channel group will eventually obsolete `oneCallMinutely`, since there is no longer a reason for keeping 60 channel groups like described here: https://www.openhab.org/addons/bindings/openweathermap/#one-call-api-minutely-forecast?

florian-h05 replied:
Renamed in https://github.com/openhab/openhab-addons/pull/15963/commits/bd7350de0c7c352b03d956bd7e3e86d53c66d229.

In https://github.com/openhab/openhab-addons/pull/15963/commits/c8c73d47010572c8908f6073eaafb52c76120d71 I have updated the handler code to always fetch hourly, daily and minutely data from the API as this is needed for the time series channels and I changed the defaults of forecastHours and forecastDays to 0, so by default only the new time series channels are created when creating a new Thing.
So we slowly get rid of the large amount of channels without breaking anything.

jlaur replied:
I'm wondering how that will actually work. I guess the defaults should also be provided here as well:
https://github.com/openhab/openhab-addons/blob/969518a077ba719d7fc5b02f461bd7b1331d91c6/bundles/org.openhab.binding.openweathermap/src/main/resources/OH-INF/config/config.xml#L132-L146

But I guess that would be a breaking change for file-based configuration when these parameters are not explicitly set?

florian-h05 replied:
The defaults are actually provided in https://github.com/openhab/openhab-addons/pull/15963/files#diff-586f2961d27e69af72e4b4e1637b6236b1c159eedbbe8c23b35c784a3ff66e95.

> But I guess that would be a breaking change for file-based configuration when these parameters are not explicitly set?

That is true, but IMHO this is no showstopper, since it is very easy to just explicitely set those in file-based configs.

jlaur replied:
> The defaults are actually provided in

Sorry, I missed that.

> > But I guess that would be a breaking change for file-based configuration when these parameters are not explicitly set?

> That is true, but IMHO this is no showstopper, since it is very easy to just explicitely set those in file-based configs.

I agree it's not a showstopper, but it will cause immediate breakage after the upgrade requiring user attention for updating configuration to get it working again. That would probably also require a note in openhab-distro, because otherwise it could be easily missed until they realize that there's no data coming in anymore?

I'm thinking if this step is one release too soon? Perhaps we could keep the existing defaults, so users would have 4.0 to transition to time series channels?

Just some thoughts for consideration before we merge. If you think it's not a big problem, we can even merge as is.

florian-h05 replied:
> That would probably also require a note in openhab-distro, because otherwise it could be easily missed until they realize that there's no data coming in anymore?

With that note and a note in the release notes, Insont see a real problem here, but …

> I'm thinking if this step is one release too soon? Perhaps we could keep the existing defaults, so users would have 4.0 to transition to time series channels?

I agree, so we can „bundle“ the default changes for the OneCall Thing with the changes for the other OWM Things and we have time to improve handling time series in rules.

florian-h05 replied:
@jlaur I have just reverted changing the defaults a few seconds ago and updated the README to no note that the defaults will change in a future release and time series usage is encouraged.

jlaur reviewed (changes requested):
Thanks for implementing time series support! LGTM, although I didn't go into full depth - I assume you already tested this as usual. 🙂 Only a few minor comments/questions.

Also, can you regenerate the I18N properties file?

lolodomo commented:
I am a little off topic in this message.

In Main UI, the only way to show these values in future will then be to look at the chart ?

We have still no way with sitemap UIs ? I believe at a certain time @J-N-K mentioned the need for a way to request chart period in the future for charts built by server. Is the chart servlet already compatible with these new time series.

jlaur commented:
> In Main UI, the only way to show these values in future will then be to look at the chart ?

Yes.

> We have still no way with sitemap UIs ? I believe at a certain time @J-N-K mentioned the need for a way to request chart period in the future for charts built by server. Is the chart servlet already compatible with these new time series ?

Sitemaps are actually still my primary way of interacting with openHAB from the Android app, but it's configured only for my production system which is still 4.0. I don't remember if I tried to configure a sitemap with a chart to inspect in BasicUI.

If you want to try it out you can simply install Energi Data Service binding. It doesn't require any authentication, and only one required parameter: Price area. Set that to one of the suggested values (e.g. DK1), and you'll get future prices for channel `spot-price`. Let me know if I can help. It seems that sitemaps in the app are quite retrospective (looking back), so maybe this was missed. Which means that actually the feature would be quite useless for my use-case, which would be viewing future prices in a sitemap chart in the app. Would be cool if anything is still possible to enhance in this area, but I guess it also requires changes in the app?

jlaur commented:
> Which means that actually the feature would be quite useless for my use-case, which would be viewing future prices in a sitemap chart in the app. Would be cool if anything is still possible to enhance in this area, but I guess it also requires changes in the app?

@lolodomo - looks not supported, this is how it looks from BasicUI:

![image](https://github.com/openhab/openhab-addons/assets/19519842/32dad62d-7ce4-47af-af4f-9c183f2f6bb5)

florian-h05 commented on bundles/org.openhab.binding.openweathermap/README.md:
```diff
@@ -268,6 +274,21 @@ The `uvindex` channel is also available in the current data and the daily foreca
+
+To configure persisting forecast data, create and link Items to those channels with time series support (as usual).
+Next, enable persistence for these Items using the `forecast` persistence strategy.
+Finally, open MainUI, search for one of the newly created Items, open the analyzer and select a future time range.
```
Addressed in https://github.com/openhab/openhab-addons/pull/15963/commits/ecb71d2788f5ef657146da6e959971fc04feb2d5 :+1:

florian-h05 commented:
@jlaur Thanks for the review.

> LGTM, although I didn't go into full depth - I assume you already tested this as usual. 

Yes, I have tested this on my dev system. I haven't tested all channels, but at least one minutely, one hourly and one daily to make sure the new code for creating time series works properly (the code to get the state is the same as before, just moved to a new private method - GitHub diff view is really bad a moved code ...). The update instructions also work fine.

> Also, can you regenerate the I18N properties file?

Done.

For the other changes I've done, please have a look at my answers to your comments.

florian-h05 commented:
> In Main UI, the only way to show these values in future will then be to look at the chart ?

Yes. Unfortunately, there it is not possible (yet) to set a chart to a future time range as default, you always have to manually go to the future.

> @lolodomo - seems not supported, this is how it looks from BasicUI, only options for looking back in time:

BasicUI would need to provide options for selecting future time ranges, but the chart servlet would also need to support future time ranges. Not sure if it does, but this should be easy to find out. Just open the browser dev tools when viewing BasicUI and find the request to the chart servlet. I would guess time range is passed as query param, so you should be able to adjust the URL and try to request a a future chart.

lolodomo commented:
The chart servlet calls the default chart provider with two dates, the start time and the end time. Here is the method in the chart provider 
 finally retrieving the data. Is this method compatible with the new feature?
https://github.com/openhab/openhab-core/blob/main/bundles/org.openhab.core.ui/src/main/java/org/openhab/core/ui/internal/chart/defaultchartprovider/DefaultChartProvider.java#L308

Thr chart servlet already supports with its calling parameters the request of a chart in future. You just have to set "begin" parameter to now and the "period" parameter as usual. So nothing to change here.

Sitemap UIs need a new input to choose between past and future. Then it is just a small change of the chart URL to be done by each UI.
The syntax of the Chart sitemap  element would require a new parameter to choose between past and future. Default would be past for backward compatibility.

lolodomo commented:
In fact, UIs should always set either begin param or end param with the current date & time.
begin -> chart in future
end -> chart in oast

lolodomo commented:
As a quick fix in Basic UI, I can add a new switch in the header row to switch between past and future.

But the user will still not be able to request directly a chart in future until we change the sitemap syntax. The parameter to add could be named "inFuture" ? Another option that is more a workaround would be to use negative periods for charts in future like for -D for the day in future.

jlaur commented:
@lolodomo - maybe a new issue should be created for adding support for future states in sitemap charts?

jlaur commented on bundles/org.openhab.binding.openweathermap/src/main/java/org/openhab/binding/openweathermap/internal/handler/OpenWeatherMapOneCallHandler.java:
```diff
@@ -75,9 +84,10 @@ public class OpenWeatherMapOneCallHandler extends AbstractOpenWeatherMapHandler
-    private int forecastHours = 24;
-    private int forecastDays = 8;
+    // forecastMinutes, -Hours and -Days determine the number of channel groups to create for each type
+    private int forecastMinutes = 0;
+    private int forecastHours = 0;
+    private int forecastDays = 0;
```
Should this be reverted as well?

florian-h05 replied:
Type of yes.
I have set this to defaults from the Thing config file, makes more sense IMO. Otherwise the Thing will rebuild its channels when reading default config while initizializing.

florian-h05 commented:
@jlaur I have addressed your review and fixed DCO (Sign-Off was missing because one of my commits this morning was from GitHub web), therefore the force push.

lolodomo commented:
Just to be clear, does this PR break something ? If a user simply upgrades OH without changing anything, will he loose some features or not ?

florian-h05 commented:
With the current implementation, this does not break anything and there will be no loss of functionality. The README just suggests/encourages to use the time series channels instead of the current channels.

jlaur reviewed (approved):
LGTM
